ludwig/models/retrieval.py

```diff
@@ -0,0 +1,225 @@
+        df = backend.batch_transform(df, self.best_batch_size, transform_fn)
+        df = backend.df_engine.compute(df)
+        embeddings = np.stack(df["data"].values).astype(np.float32)
+        return embeddings
+
+    def search(
+        self, df: pd.DataFrame, backend: "Backend", k: int = 10, return_data: bool = False
+    ) -> Union[List[int], List[Dict[str, Any]]]:
+        rows = df.to_dict(orient="records")
+        row_strs = [json.dumps(r) for r in rows]
+
+        query_vectors = self._encode(row_strs, backend)
+        results = []
+        # TODO(geoffrey): figure out why self.index.search segfaults with larger batch sizes
+        for query_vector in tqdm(query_vectors, total=query_vectors.shape[0]):
+            top_k = self.index.search(query_vector.reshape(1, -1), k)
+            indices = top_k[1].tolist()[0]
+
+            if return_data:
+                result = self.index_data.iloc[indices].to_dict(orient="records")
+            else:
+                result = indices
+            results.append(result)
+        return results
+
+    def save_index(self, name: str, cache_directory: str):
+        index_file_path = os.path.join(cache_directory, name + ".index")
+        faiss.write_index(self.index, index_file_path)
+
+        index_data_file_path = os.path.join(cache_directory, name + "_data.csv")
+        self.index_data.to_csv(index_data_file_path, index=False)
+
+    def load_index(self, name: str, cache_directory: str):
+        index_file_path = os.path.join(cache_directory, name + ".index")
+        self.index = faiss.read_index(index_file_path)
+
+        index_data_file_path = os.path.join(cache_directory, name + "_data.csv")
+        self.index_data = pd.read_csv(index_data_file_path)
+
+
+def create_semantic_retrieval_model_evaluator(model_name, samples):
+    class _RetrievalModelEvaluator(BatchSizeEvaluator):
+        def __init__(self):
+            self.model = get_semantic_retrieval_model(model_name)
+            self.samples = samples
+
+        def step(self, batch_size: int):
+            self.model.encode(self.samples[:batch_size], batch_size=batch_size, show_progress_bar=False)
+
+    return _RetrievalModelEvaluator
+
+
+def create_semantic_retrieval_model_fn(model_name, batch_size):
+    class _RetrievalModelFn:
+        def __init__(self):
+            self.model = get_semantic_retrieval_model(model_name)
+            self.batch_size = batch_size
+
+        def __call__(self, df: pd.DataFrame) -> np.ndarray:
+            row_strs = df["data"].tolist()
+            result = self.model.encode(row_strs, batch_size=self.batch_size, show_progress_bar=False)
+            df["data"] = result.tolist()
+            return df
+
+    return _RetrievalModelFn
+
+
+def get_semantic_retrieval_model(model_name):
+    from sentence_transformers import SentenceTransformer
+
+    return SentenceTransformer(model_name, device=get_torch_device())
```

Comment:
OOC, in multi-gpu setups, where does the semantic retrieval model get placed with the call to `get_torch_device()` since say it will return `cuda`. Does it place it on the CPU if no GPUs are available, or the first GPU device available?

Reply:
It should be placed on whatever device is visible to Ray Datasets, so in theory it should be placed on the next free GPU (if GPU resources were requested by the job)